README.md
![Dynamic JSON Badge](https://img.shields.io/badge/dynamic/json?url=https%3A%2F%2Fraw.githubusercontent.com%2Fstrawmelonjuice%2Fhl-img%2Fmain%2Fpackage.json&query=%24.version&label=Version&labelColor=%23313333&color=%23eebcbc&link=https%3A%2F%2Fgithub.com%2Fstrawmelonjuice%2Fhl-img&link=https%3A%2F%2Fwww.npmjs.com%2Fpackage%2Fhl-img%3FactiveTab%3Dversions) [![license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/strawmelonjuice/hl-img/blob/main/LICENSE)
[![](https://data.JSdelivr.com/v1/package/npm/hl-img/badge)](https://www.JSdelivr.com/package/npm/hl-img)

# hl-img

I added a custom element into my site for highlightable images, now I need to make it a repo so I don't lose track.
## NOTE:
> Google does not like the use of elements like these, mainly because browsers or bots that don't fire the javascript, will not be able to see any images. _Though Google's HTML-checker web.dev loads `hl-img` in their tests, allowing it to pass that test after all._

## How to embed

Thanks to the `defer` attribute, they'll load after the body is loaded, so you _can_ place them anywhere. However, for best reliability, please add this at the end of the `<body>`.

### From JSdelivr
_Minified_
```html
<script id="hlimg-options" type="application/json">{"timing_fade_opacity": 0.5,	"timing_pop_location": 700,	"styling_closenotif_positioning": "top: 2em; right: 2em;",	"styling_closenotif_colors_background": "#000",	"styling_closenotif_colors_text": "#FFF",	"styling_closenotif_border_thickness": "2.5px",	"styling_closenotif_border_type": "solid","styling_closenotif_border_color": "#848484","styling_closenotif_padding": "4px","styling_imageshow_zIndex": 900,"styling_imageshow_positioning": "top: 20%; bottom: 20%; left: 10%; right: 10%;","styling_imageshow_colors_background": "#f5f5f5","styling_imageshow_border_thickness": ".2em","styling_imageshow_border_type": "solid","styling_imageshow_border_color": "currentColor","styling_imageshow_border_radius": "0%","styling_hlimg_maxwidth": "80%","styling_hlimg_maxheight": "100%"}</script>
<script defer type="module" src="https://cdn.jsdelivr.net/npm/hl-img@2/dist/hl-img.min.js"></script>
```
For the extra-minified version, just use `https://cdn.jsdelivr.net/npm/hl-img@2/dist/hl-img.e.min.js`, instead.

_Unminified_

```html
<script id="hlimg-options" type="application/json">{
	"timing_fade_opacity": 0.5,
	"timing_pop_location": 700,
	"styling_closenotif_positioning": "top: 2em; right: 2em;",
	"styling_closenotif_colors_background": "#000",
	"styling_closenotif_colors_text": "#FFF",
	"styling_closenotif_border_thickness": "2.5px",
	"styling_closenotif_border_type": "solid",
	"styling_closenotif_border_color": "#848484",
	"styling_closenotif_padding": "4px",
	"styling_imageshow_zIndex": 900,
	"styling_imageshow_positioning": "top: 20%; bottom: 20%; left: 10%; right: 10%;",
	"styling_imageshow_colors_background": "#f5f5f5",
	"styling_imageshow_border_thickness": ".2em",
	"styling_imageshow_border_type": "solid",
	"styling_imageshow_border_color": "currentColor",
	"styling_imageshow_border_radius": "0%",
	"styling_hlimg_maxwidth": "80%",
	"styling_hlimg_maxheight": "100%"
}</script>
<script defer type="module" src="https://cdn.jsdelivr.net/npm/hl-img@2/dist/hl-img.js"></script>
```

### Version-specific

For version-specific JSdelivr's, browse <https://cdn.jsdelivr.net/npm/hl-img/>.

NOTE: Versions before v2 worked really different. See the [old readme](https://github.com/strawmelonjuice/hl-img/blob/7def3ff19eae6368f18c0f7c38d44a4706d08170/README.md) for those.

## Install using NPM
```bash
npm i hl-img
```
And then embed it from your own host.

```html
<script id="hlimg-options" type="application/json">{
	"timing_fade_opacity": 0.5,
	"timing_pop_location": 700,
	"styling_closenotif_positioning": "top: 2em; right: 2em;",
	"styling_closenotif_colors_background": "#000",
	"styling_closenotif_colors_text": "#FFF",
	"styling_closenotif_border_thickness": "2.5px",
	"styling_closenotif_border_type": "solid",
	"styling_closenotif_border_color": "#848484",
	"styling_closenotif_padding": "4px",
	"styling_imageshow_zIndex": 900,
	"styling_imageshow_positioning": "top: 20%; bottom: 20%; left: 10%; right: 10%;",
	"styling_imageshow_colors_background": "#f5f5f5",
	"styling_imageshow_border_thickness": ".2em",
	"styling_imageshow_border_type": "solid",
	"styling_imageshow_border_color": "currentColor",
	"styling_imageshow_border_radius": "0%",
	"styling_hlimg_maxwidth": "80%",
	"styling_hlimg_maxheight": "100%"
}</script>
<script defer type="module" src="someplace/node_modules/hl-img/dist/hl-img.js"></script>
```




## Attributes
### An example
```html
<hl-img preview="/assets/img/media/art/MM2.png" alt="MouseMouse!" highlight="/assets/img/media/art/MM2.webp"></hl-img>
```
[This](https://strawmelonjuice.com/blog?p=posts/art/mousemouse-3.2) was my first use case of `<hl-img>`.
### Supported
- `preview` – The `preview`-attribute is used as the `src` attribute when the image is _not_ in highlight-mode. If not specified, will be inherited from `src`.

- `highlight` – The `highlight`-attribute is used as the `src` attribute when the image is currently in highlight-mode. If not specified, will be inherited from `preview`.

- `href` – `href` attributes specify which link to open on clicking on the highlighted image.

### Inherited
- `alt` – The `<img>`'s `alt`-attribute is copied one-on-one. No changes in functioning.

- `src` – `<hl-img>`-elements do support `<img>`'s `src`-attribute, but will only use it as a fallback, in case `preview` is not specified.

- `style` – Styles are inherited, but may be altered by `hl-img`.

- `id` and `class` – Ids and classes are inherited, however, if none are specified, `hl-img` will assign them. It will also add the `hl-image` class in all cases.

- `width` and `height` – Are inherited.

### Future support

Add an issue if you're missing anything.

## Customization

Before calling `hl-img`, you can declare some of it's options, if you don't, their defaults will be used.

An example of how an custom configuration would look in HTML:

```HTML
<html>
    <head>
        ...
    </head>
<body>
   ...
    <main>
        ...
    </main>
<script id="hlimg-options" type="application/json">
    {"styling_closenotif_positioning":"bottom: 2em; right: 4em;","styling_closenotif_colors_background":"#FFF","styling_closenotif_colors_text":"#000","styling_imageshow_zIndex":900000,"styling_imageshow_colors_background":"#000","styling_imageshow_border_radius":"25%"}
</script>
<script defer type="module" src="hl-img"></script>
</body>
</html>
```

Note that these changes happen globally, so they'll be applied to the entire page, the moment `hl-img` is ran.

### List of customization variables

- Timing:

  - `timing_fade_opacity`  (number) – The time in seconds of the fade-in/fade-out animation on the images. Default: `0.5`

  - `timing_pop_location` (number)  – Time in miliseconds to wait before moving images. By default a little longer than `timing_fade_opacity` so that it can run its animation. Default: `700`

- Styling

  - Closing modal

    - `styling_closenotif_positioning` (string) – CSS code for positioning the closing modal. Default: `"top: 2em; right: 2em"`

    - `styling_closenotif_colors_background` (string) – Background colour for the closing modal. Default: `"#000"`

    - `styling_closenotif_colors_text` (string) – Text colour for the closing modal. Default: `"#FFF"`

    - `styling_closenotif_border_thickness` (string) – Border thickness for the closing modal. Default: `"2.5px"`

    - `styling_closenotif_border_type` (string) – Type of the closing modal's border, CSS value. Default: `"solid"`

    - `styling_closenotif_border_color`  (string) –  The colour of the closing modal's border. Default: `"#848484"`

    - `styling_closenotif_padding` (string) – Padding for the text inside the closing modal. Default: `"4px"`

  - Highlighted image

    - `styling_imageshow_zIndex` (number) – Z-Index for the image when highlighted, the Z-Index of the closing modal will always be this +1. CSS value. Default: `900`

    -   `styling_imageshow_positioning` (string) – CSS code for positioning the highlighted image. Default: `top: 20%; bottom: 20%; left: 10%; right: 10%"`

    - `styling_imageshow_colors_background` (string) – Background colour for highlighted image -- shows when image is transparent. Default: `"#f5f5f5"`

    - `styling_imageshow_border_thickness` (string) – Border thickness for the highlighted image. Default: `".2em"`
    - `styling_imageshow_border_type` (string) – Type of the highlighted images' border, CSS value. Default: `"solid"`
    - `styling_imageshow_border_color` (string) – Highlighted images' border color. Default:`"currentColor"`
    - `styling_imageshow_border_radius` (string) – Border radius for highlighted image. Default:`"0%"`
    - `styling_hlimg_maxwidth` (string) – Max-width for all `<hlimg>`'s. Set to `""` To disable. Default: `"80%"`
    - `styling_hlimg_maxwidth` (string) – Max-height for all `<hlimg>`'s. Set to `""` To disable. Default: `"100%"`


CHANGELOG.MD
# Changelog

## Unreleased

-  **<span style="color: #0033cc">Update</span>**: Updated demo version to v2+
-  **<span style="color: #0033cc">Update</span>**: minified inline styling.



## v2.0.1 release

- <span style="color: red; background-color: #ffcc00">BREAKING!</span> **<span style="color: #0033cc">Update</span>**: We changed over to Typescript and are now a module!





<hr>

This file was auto generated by [<span style="background-color: #24273a; color: #ffcc00">Bananen! 🍌</span>](https://github.com/strawmelonjuice/bananen/) `v0.0.3-alpha`.

test/js/1.js
// // 3rd file to run.
setTimeout(function () {
    console.log("1.js is the 3rd file to run");
    console.log("latest: " + window.parent.hlimgversions.latest + ", local: " + window.parent.hlimgversions.local)
    document.getElementById("latestver").innerHTML = window.parent.hlimgversions.latest;
    document.getElementById("localver").innerHTML = window.parent.hlimgversions.local;
}, 1500);


test/js/2.js
// 1st file to run
setTimeout(function () {
    console.log("2.js is the 1st file to run");
    try {
        window.parent.hlimgversions = {}
    } catch (TypeError) {
        alert("Please run this testing page from a localhost http server for the best results.");
        stop();
    }
    try {
        output = document.getElementById('hl-img-version').value;
    } catch (TypeError) {
        output = "1.10.4 or older."
    }
    output = "v" + output;
    window.parent.hlimgversions.local = output;
}, 500);


test/js/3.js
// 2nd file to run
setTimeout(function () {
    console.log("3.js is the 2nd file to run");
    try {
        output = document.getElementById('hl-img-version').value;        
    } catch (TypeError) {
        if (ver !== null) {
            output = ver;
        } else if (hlimg_ver !== null) {
            output = hlimg_ver;
        } else {
            output = "1.10.4 or older."
        }
    }
    output = "v" + output;
    window.parent.hlimgversions.latest = output;
}, 1000);


hl-img.js
const hlimg_ver = "1.10.11";
let hlimg_opt;
const hlimg_options_default = {
  timing_fade_opacity: 0.5,
  timing_pop_location: 700,
  styling_closenotif_positioning: "top: 2em; right: 2em;",
  styling_closenotif_colors_background: "#000",
  styling_closenotif_colors_text: "#FFF",
  styling_closenotif_border_thickness: "2.5px",
  styling_closenotif_border_type: "solid",
  styling_closenotif_border_color: "#848484",
  styling_closenotif_padding: "4px",
  styling_imageshow_zIndex: 900,
  styling_imageshow_positioning:
    "top: 20%; bottom: 20%; left: 10%; right: 10%;",
  styling_imageshow_colors_background: "#f5f5f5",
  styling_imageshow_border_thickness: ".2em",
  styling_imageshow_border_type: "solid",
  styling_imageshow_border_color: "currentColor",
  styling_imageshow_border_radius: "0%",
  styling_hlimg_maxwidth: "80%",
  styling_hlimg_maxheight: "100%",
};
if (!(typeof hlimg_options === "undefined")) {
  hlimg_opt = Object.assign(hlimg_options_default, hlimg_options);
} else {
  hlimg_opt = hlimg_options_default;
}

const css = `
  img.hl-img {
      -webkit-transition: opacity ${hlimg_opt.timing_fade_opacity}s ease-in-out;
      -moz-transition: opacity ${hlimg_opt.timing_fade_opacity}s ease-in-out;
      transition: opacity ${hlimg_opt.timing_fade_opacity}s ease-in-out;
  }
  .hl-img:not(.imageshow) {
    cursor: zoom-in;
    object-fit: contain;
  }
  #closenotifelm {
      position: fixed;
      ${hlimg_opt.styling_closenotif_positioning};
      background-color: ${hlimg_opt.styling_closenotif_colors_background};
      color: ${hlimg_opt.styling_closenotif_colors_text};
      padding: ${hlimg_opt.styling_closenotif_padding};
      border: ${hlimg_opt.styling_closenotif_border_thickness} ${
  hlimg_opt.styling_closenotif_border_type
} ${hlimg_opt.styling_closenotif_border_color};
      z-index: ${hlimg_opt.styling_imageshow_zIndex + 1};
  }
  img.imageshow {
      position: fixed;
      margin: auto;
      ${hlimg_opt.styling_imageshow_positioning};
      z-index: ${hlimg_opt.styling_imageshow_zIndex};
      background-color: ${hlimg_opt.styling_imageshow_colors_background};
      border-radius: ${hlimg_opt.styling_imageshow_border_radius};
      border: ${hlimg_opt.styling_imageshow_border_color} ${
  hlimg_opt.styling_imageshow_border_type
} ${hlimg_opt.styling_imageshow_border_thickness};
      box-shadow: -53px -38px 95px 0px rgba(0, 0, 0, 0.86);
        -webkit-box-shadow: -53px -38px 95px 0px rgba(0, 0, 0, 0.86);
        -moz-box-shadow: -53px -38px 95px 0px rgba(0, 0, 0, 0.86);
  }
`;

const styleSheet = document.createElement("style");
styleSheet.innerHTML = css;
document.head.appendChild(styleSheet);

const versionDec = document.createElement("data");
versionDec.id = "hl-img-version";
versionDec.value = hlimg_ver;
document.head.appendChild(versionDec);

const imoges = document.getElementsByTagName("hl-img");
for (i = imoges.length - 1; i >= 0; i--) {
  const newimogesid =
    `hl-img-${Math.floor(
      Math.random() * (Math.floor(9999) - Math.ceil(1000) + 1) + Math.ceil(1000)
    )}`;
  const newimoges = document.createElement("img");
  newimoges.alt = imoges.item(i).getAttribute("alt");
  if (imoges.item(i).getAttribute("style") != null) {
    newimoges.setAttribute("style", imoges.item(i).getAttribute("style"));
  }
  if (imoges.item(i).getAttribute("height") != null) {
    newimoges.setAttribute("height", imoges.item(i).getAttribute("height"));
    newimoges.dataset.htmlheight = imoges.item(i).getAttribute("height");
  }
  if (imoges.item(i).getAttribute("width") != null) {
    newimoges.setAttribute("width", imoges.item(i).getAttribute("width"));
    newimoges.dataset.htmlwidth = imoges.item(i).getAttribute("width");
  }
  // I make this typo a lot lmao
  if (imoges.item(i).getAttribute("widht") != null) {
    newimoges.setAttribute("width", imoges.item(i).getAttribute("widht"));
    newimoges.dataset.htmlwidth = imoges.item(i).getAttribute("widht");
  }
  if (imoges.item(i).style.height != null) {
    newimoges.setAttribute("height", imoges.item(i).style.height);
    newimoges.dataset.htmlheight = imoges.item(i).style.height;
  }
  if (imoges.item(i).style.width != null) {
    newimoges.setAttribute("width", imoges.item(i).style.width);
    newimoges.dataset.htmlwidth = imoges.item(i).style.width;
  }
  if (imoges.item(i).getAttribute("class") != null) {
    newimoges.setAttribute("class", imoges.item(i).getAttribute("class"));
  }
  if (imoges.item(i).getAttribute("id") != null) {
    newimoges.setAttribute("id", imoges.item(i).getAttribute("id"));
  } else {
    newimoges.id = newimogesid;
  }
  newimoges.setAttribute("onclick", "imageshow(1, this)");
  if (imoges.item(i).getAttribute("preview") != null) {
    newimoges.dataset.prevsrc = imoges.item(i).getAttribute("preview");
    newimoges.src = imoges.item(i).getAttribute("preview");
  } else {
    newimoges.src = imoges.item(i).getAttribute("src");
    newimoges.dataset.prevsrc = imoges.item(i).getAttribute("src");
  }
  if (imoges.item(i).getAttribute("highlight") != null) {
    newimoges.dataset.showsrc = imoges.item(i).getAttribute("highlight");
  } else {
    newimoges.dataset.showsrc = newimoges.dataset.prevsrc;
  }
  if (imoges.item(i).getAttribute("href") != null) {
    newimoges.dataset.href = imoges.item(i).getAttribute("href");
  }
  newimoges.style.maxWidth = hlimg_opt.styling_hlimg_maxwidth;
  newimoges.style.maxHeight = hlimg_opt.styling_hlimg_maxheight;
  newimoges.title = "Click to show image";
  newimoges.classList.add("hl-img");
  imoges.item(i).parentNode.replaceChild(newimoges, imoges.item(i));
}

if (document.getElementsByClassName("hl-img")[0]) {
  const closenotif = document.createElement("div");
  closenotif.innerHTML =
    "<p>Click outside of the image to put it back in place!</p>";
  closenotif.id = "closenotifelm";
  closenotif.style.display = "none";
  document.body.appendChild(closenotif);
}
closenotif = document.getElementById("closenotifelm");
function imageshow(action, elem) {
  if (action) {
    console.log(`1/2 Zooming in to ${elem.id}`);
    elem.style.opacity = "0%";
    setTimeout(function () {
      console.log(`2/2 Zooming in to ${elem.id}`);
      elem.classList.add("imageshow");
      elem.setAttribute("onclick", "");
      if (elem.dataset.href != null) {
        elem.setAttribute(
          "onclick",
          `window.open('${elem.dataset.href}');`
        );
        closenotif.innerHTML =
          `<p>Click outside of the image to put it back in place! Or click on the image again to open <a href="${elem.dataset.href}" target="_blank">this link</a> in a new tab.</p>`;
        elem.style.cursor = "pointer";
      } else {
        closenotif.innerHTML =
          "<p>Click outside of the image to put it back in place!</p>";
      }
      elem.style.height = "";
      elem.setAttribute("height", "");
      elem.style.width = "";
      elem.setAttribute("width", "");
      elem.src = elem.dataset.showsrc;
      elem.style.opacity = "100%";
      elem.title = "";
      closenotif.style.display = "block";
    }, hlimg_opt.timing_pop_location);
    const abortController = new AbortController();
    window.addEventListener(
      "click",
      function (e) {
        if (!elem.contains(e.target)) {
          imageshow(0, elem);
          abortController.abort();
        }
      },
      { signal: abortController.signal }
    );
  } else {
    console.log(`1/2 Zooming out of ${elem.id}`);
    elem.style.opacity = "0%";
    setTimeout(function () {
      console.log(`2/2 Zooming out of ${elem.id}`);
      elem.classList.remove("imageshow");
      elem.setAttribute("onclick", "imageshow(1, this)");
      elem.style.cursor = "";
      if (elem.dataset.htmlheight != null) {
        elem.style.height = elem.dataset.htmlheight;
        elem.setAttribute("height", elem.dataset.htmlheight);
      }
      if (elem.dataset.htmlwidth != null) {
        elem.style.width = elem.dataset.htmlwidth;
        elem.setAttribute("width", elem.dataset.htmlwidth);
      }
      elem.src = elem.dataset.prevsrc;
      elem.style.opacity = "100%";
      elem.title = "Click to show image";
      closenotif.style.display = "none";
    }, hlimg_opt.timing_pop_location);
  }
}
console.info(
  `%c hl-img.js by Mar was loaded!%c\nVersion: ${hlimg_ver}\n\r\n for more info check it out on GitHub:\n%chttps://github.com/strawmelonjuice/hl-img.js/`,
  "color: #8A2BE2; font-size:15px;",
  "color: black; font-size:12px;",
  "color: blue; font-size:12px;"
);


src/hl-img.ts
if (typeof window.document === "undefined") {
	throw new Error("HL-IMG is meant for usage in the browser!");
}
const hlimg_ver = "2.0.2";
if (window.document.getElementsByClassName("hl-img")[0] !== null) {
	const closenotif = window.document.createElement("div");
	closenotif.innerHTML =
		"<p>Click outside of the image to put it back in place!</p>";
	closenotif.id = "closenotifelm";
	closenotif.style.display = "none";
	window.document.body.appendChild(closenotif);
	console.log("Creating close notification");
}
const closenotif = window.document.getElementById("closenotifelm");
let hlimg_opt: {
	timing_fade_opacity: number;
	styling_closenotif_positioning: string;
	styling_closenotif_colors_background: string;
	styling_closenotif_colors_text: string;
	styling_closenotif_padding: string;
	styling_closenotif_border_thickness: string;
	styling_closenotif_border_type: string;
	styling_closenotif_border_color: string;
	styling_imageshow_zIndex: number;
	styling_imageshow_positioning: string;
	styling_imageshow_colors_background: string;
	styling_imageshow_border_radius: string;
	styling_imageshow_border_color: string;
	styling_imageshow_border_type: string;
	styling_imageshow_border_thickness: string;
	styling_hlimg_maxwidth: string;
	styling_hlimg_maxheight: string;
	timing_pop_location: number;
};
const hlimg_options_default = {
	timing_fade_opacity: 0.5,
	timing_pop_location: 700,
	styling_closenotif_positioning: "top: 2em; right: 2em;",
	styling_closenotif_colors_background: "#000",
	styling_closenotif_colors_text: "#FFF",
	styling_closenotif_border_thickness: "2.5px",
	styling_closenotif_border_type: "solid",
	styling_closenotif_border_color: "#848484",
	styling_closenotif_padding: "4px",
	styling_imageshow_zIndex: 900,
	styling_imageshow_positioning:
		"top: 20%; bottom: 20%; left: 10%; right: 10%;",
	styling_imageshow_colors_background: "#f5f5f5",
	styling_imageshow_border_thickness: ".2em",
	styling_imageshow_border_type: "solid",
	styling_imageshow_border_color: "currentColor",
	styling_imageshow_border_radius: "0%",
	styling_hlimg_maxwidth: "80%",
	styling_hlimg_maxheight: "100%",
};

let hlimg_options;
if (window.document.getElementById("hlimg-options") !== null) {
	hlimg_options = JSON.parse(
		window.document.getElementById("hlimg-options").textContent,
	);
}
if (!(typeof hlimg_options === "undefined")) {
	hlimg_opt = Object.assign(hlimg_options_default, hlimg_options);
} else {
	hlimg_opt = hlimg_options_default;
}
let vierdriedrievijf = hlimg_opt.styling_imageshow_zIndex;
vierdriedrievijf += 1;
const css = `
img.hl-img{-webkit-transition:opacity ${hlimg_opt.timing_fade_opacity}s ease-in-out;-moz-transition:opacity ${hlimg_opt.timing_fade_opacity}s ease-in-out;transition:opacity ${hlimg_opt.timing_fade_opacity}s ease-in-out;}.hl-img:not(.imageshow){cursor:zoom-in;object-fit:contain}#closenotifelm{position:fixed;${hlimg_opt.styling_closenotif_positioning};background-color:${hlimg_opt.styling_closenotif_colors_background};color:${hlimg_opt.styling_closenotif_colors_text};padding:${hlimg_opt.styling_closenotif_padding};border:${hlimg_opt.styling_closenotif_border_thickness}${hlimg_opt.styling_closenotif_border_type}${hlimg_opt.styling_closenotif_border_color};z-index:${vierdriedrievijf};}img.imageshow{position:fixed;margin:auto;${hlimg_opt.styling_imageshow_positioning};z-index:${hlimg_opt.styling_imageshow_zIndex};background-color:${hlimg_opt.styling_imageshow_colors_background};border-radius:${hlimg_opt.styling_imageshow_border_radius};border:${hlimg_opt.styling_imageshow_border_color} ${hlimg_opt.styling_imageshow_border_type} ${hlimg_opt.styling_imageshow_border_thickness};box-shadow:-53px -38px 95px 0px rgba(0,0,0,0.86);-webkit-box-shadow:-53px -38px 95px 0px rgba(0,0,0,0.86);-moz-box-shadow:-53px -38px 95px 0px rgba(0,0,0,0.86);}
`;

const styleSheet = window.document.createElement("style");
styleSheet.innerHTML = css;
window.document.head.appendChild(styleSheet);

const versionDec = window.document.createElement("data");
versionDec.id = "hl-img-version";
versionDec.value = hlimg_ver;
window.document.head.appendChild(versionDec);

const imoges = window.document.getElementsByTagName(
	"hl-img",
) as HTMLCollectionOf<HTMLElement>;
for (let i = imoges.length - 1; i >= 0; i--) {
	const newimogesid = `hl-img-${Math.floor(
		Math.random() * (Math.floor(9999) - Math.ceil(1000) + 1) + Math.ceil(1000),
	)}`;
	const newimoges = window.document.createElement("img");
	newimoges.alt = imoges.item(i).getAttribute("alt");
	if (imoges !== null) {
		if (imoges.item(i).getAttribute("style") != null) {
			newimoges.setAttribute("style", imoges.item(i).getAttribute("style"));
		}
		if (imoges.item(i).getAttribute("height") != null) {
			newimoges.setAttribute("height", imoges.item(i).getAttribute("height"));
			newimoges.dataset.htmlheight = imoges.item(i).getAttribute("height");
		}
		if (imoges.item(i).getAttribute("width") != null) {
			newimoges.setAttribute("width", imoges.item(i).getAttribute("width"));
			newimoges.dataset.htmlwidth = imoges.item(i).getAttribute("width");
		}
		// I make this typo a lot lmao
		if (imoges.item(i).getAttribute("widht") != null) {
			newimoges.setAttribute("width", imoges.item(i).getAttribute("widht"));
			newimoges.dataset.htmlwidth = imoges.item(i).getAttribute("widht");
		}
		if (imoges.item(i).style.height != null) {
			newimoges.setAttribute("height", imoges.item(i).style.height);
			newimoges.dataset.htmlheight = imoges.item(i).style.height;
		}
		if (imoges.item(i).style.width != null) {
			newimoges.setAttribute("width", imoges.item(i).style.width);
			newimoges.dataset.htmlwidth = imoges.item(i).style.width;
		}
		if (imoges.item(i).getAttribute("class") != null) {
			newimoges.setAttribute("class", imoges.item(i).getAttribute("class"));
		}
		if (imoges.item(i).getAttribute("id") != null) {
			newimoges.setAttribute("id", imoges.item(i).getAttribute("id"));
		} else {
			newimoges.id = newimogesid;
		}

		newimoges.setAttribute("onclick", "this.imageshow(1, this)");
		newimoges.imageshow = function (
			action: boolean,
			elem: {
				id: string;
				style: {
					opacity: string;
					cursor: string;
					height: string;
					width: string;
				};
				classList: {
					add: (arg0: string) => void;
					remove: (arg0: string) => void;
				};
				setAttribute: (arg0: string, arg1: string) => void;
				dataset: {
					href: null;
					showsrc: string;
					htmlheight: null;
					htmlwidth: null;
					prevsrc: string;
				};
				src: string;
				title: string;
				contains: (arg0: string) => string;
			},
		) {
			if (action) {
				console.log(`1/2 Zooming in to ${elem.id}`);
				elem.style.opacity = "0%";
				setTimeout(function () {
					console.log(`2/2 Zooming in to ${elem.id}`);
					elem.classList.add("imageshow");
					elem.setAttribute("onclick", "");
					if (elem.dataset.href != null) {
						elem.setAttribute(
							"onclick",
							`window.open('${elem.dataset.href}');`,
						);
						closenotif.innerHTML = `<p>Click outside of the image to put it back in place! Or click on the image again to open <a href="${elem.dataset.href}" target="_blank">this link</a> in a new tab.</p>`;
						elem.style.cursor = "pointer";
					} else {
						closenotif.innerHTML =
							"<p>Click outside of the image to put it back in place!</p>";
					}
					elem.style.height = "";
					elem.setAttribute("height", "");
					elem.style.width = "";
					elem.setAttribute("width", "");
					elem.src = elem.dataset.showsrc;
					elem.style.opacity = "100%";
					elem.title = "";
					closenotif.style.display = "block";
				}, hlimg_opt.timing_pop_location);
				const abortController = new AbortController();
				window.addEventListener(
					"click",
					// rome-ignore lint/suspicious/noExplicitAny: <explanation>
					function (e: { target: HTMLImageElement }) {
						if (!elem.contains(e.target)) {
							elem.imageshow(0, elem);
							abortController.abort();
						}
					},
					{ signal: abortController.signal },
				);
			} else {
				console.log(`1/2 Zooming out of ${elem.id}`);
				elem.style.opacity = "0%";
				setTimeout(function () {
					console.log(`2/2 Zooming out of ${elem.id}`);
					elem.classList.remove("imageshow");
					elem.setAttribute("onclick", "this.imageshow(1, this)");
					elem.style.cursor = "";
					if (elem.dataset.htmlheight != null) {
						elem.style.height = elem.dataset.htmlheight;
						elem.setAttribute("height", elem.dataset.htmlheight);
					}
					if (elem.dataset.htmlwidth != null) {
						elem.style.width = elem.dataset.htmlwidth;
						elem.setAttribute("width", elem.dataset.htmlwidth);
					}
					elem.src = elem.dataset.prevsrc;
					elem.style.opacity = "100%";
					elem.title = "Click to show image";
					closenotif.style.display = "none";
				}, hlimg_opt.timing_pop_location);
			}
		};
		if (imoges.item(i).getAttribute("preview") != null) {
			newimoges.dataset.prevsrc = imoges.item(i).getAttribute("preview");
			newimoges.src = imoges.item(i).getAttribute("preview");
		} else {
			newimoges.src = imoges.item(i).getAttribute("src");
			newimoges.dataset.prevsrc = imoges.item(i).getAttribute("src");
		}
		if (imoges.item(i).getAttribute("highlight") != null) {
			newimoges.dataset.showsrc = imoges.item(i).getAttribute("highlight");
		} else {
			newimoges.dataset.showsrc = newimoges.dataset.prevsrc;
		}
		if (imoges.item(i).getAttribute("href") != null) {
			newimoges.dataset.href = imoges.item(i).getAttribute("href");
		}
		newimoges.style.maxWidth = hlimg_opt.styling_hlimg_maxwidth;
		newimoges.style.maxHeight = hlimg_opt.styling_hlimg_maxheight;
		newimoges.title = "Click to show image";
		newimoges.classList.add("hl-img");
		imoges.item(i).parentNode.replaceChild(newimoges, imoges.item(i));
	}
}

console.info(
	`%c hl-img.js by Mar was loaded!%c\nVersion: ${hlimg_ver}\n\r\n for more info check it out on GitHub:\n%chttps://github.com/strawmelonjuice/hl-img.js/`,
	"color: #8A2BE2; font-size:15px;",
	"color: black; font-size:12px;",
	"color: blue; font-size:12px;",
);
